### ISSUE TITLE
Getting warning 'video is being deferred until the player has loaded'

### ISSUE THREAD

**Issue by djohal** [CookPete/react-player#413]
I'm getting this warning every time player is trying to load my video. It works fine when it loads the video for the first time, but after that for all other videos it shows a glimpse of previous video and then loads the actual one.

How can I wait for the player to fully load before sending in the url?

This is my what my code looks like:
```
<ReactPlayer  
    playing={false} 
    controls={true}
    url={this.props.movieVid}
/>
```

**Comment by CookPete** [comment]
The warning happens [here](https://github.com/CookPete/react-player/blob/9618272e6e4b518e493d3130b96dd2ebeb97b506/src/Player.js#L37) when a new `url` is passed in _before_ the `onReady` or `onPlay` event occurs. This behaviour was added to fix https://github.com/CookPete/react-player/issues/377.

Is this what is happening? What URLs are you trying to play?

**Comment by djohal** [comment]
I'm passing in youtube video that I'm getting from prop. How do I pass the url after onReady event?

**Comment by CookPete** [comment]
>How do I pass the url after onReady event?

You should be just passing the `url` in on first render, not trying to pass it in at a later time. `onReady` is just a callback for when the player deems itself "ready" to play something. If the `url` you are passing in _changes_ between the first render of ReactPlayer and the `onReady` event, you will get this warning. ReactPlayer will then attempt to load the `url` you changed to once the player is ready, which may show a flick of a previous video.

**Comment by ryanyu104** [comment]
@djohal How to pass the URLs on every render?  I also getting URLs from prop. THX

**Comment by dChunikhin** [comment]
@djohal could you please describe, where did you add conditional statement to resolve the issue? Thanks in advance!

**Comment by rkshtsingh28** [comment]
@cookpete can you please help me out with same issue 'warning - video is being deferred until the player has loaded', i didn't get the solution clearly.

**Comment by cookpete** [comment]
The solution is to not change the `url` prop too quickly. This happens when a player tries to load but `url` changes before the player is ready, as described in [my previous comment](https://github.com/cookpete/react-player/issues/413#issuecomment-394468325) from nearly 3 years ago.

**Comment by MitchelSt** [comment]
@cookpete what would be the preferred approach in case you do want to dynamically change the url on, for example, a click event elsewhere in the application?

**Comment by cookpete** [comment]
>what would be the preferred approach in case you do want to dynamically change the url on, for example, a click event elsewhere in the application?

There are many answers for this (and it's not really relevant to the issue) but the most basic way would be to have something like

```js
const [url, setUrl] = useState(null)
```

in a top-level component somewhere. Then pass `url` down to your video player component that does 

```jsx
<ReactPlayer={url} ... />
``` 

and then pass `setUrl` down to the component you want to change the URL, then just do something like

```js
<button onClick={() => setUrl('https://whatever')}>Click me!</button>
```

You could also use [context](https://reactjs.org/docs/context.html), or [redux](https://react-redux.js.org/), or any state management library.

**Comment by MitchelSt** [comment]
> You could also use [context](https://reactjs.org/docs/context.html), or [redux](https://react-redux.js.org/), or any state management library.

The reason I asked this question right here is because I, like you said, changed the `url` dynamically (in my case it's essentially a useState hook that drills down the `url` as a prop).

However, when I run tests (with the React Testing Library), I get this warning in the console.

**Comment by cookpete** [comment]
Ah yeah, changing `url` rapidly (like in a test) is likely to trigger that warning. That’s annoying.

I'm considering getting rid of that guard altogether, or at least a prop that avoids it. It seems to cause more problems than it solves.

**Comment by plantshopping** [comment]
Also having the same problem as @MitchelSt

**Comment by kusgupta** [comment]
Also having the same problem as @MitchelSt as well.
